addons/website/static/src/js/editor/snippets.options.js

```diff
@@ -316,66 +370,194 @@ const FontFamilyPickerUserValueWidget = SelectUserValueWidget.extend({
+            }
+            async onUploadChange(e) {
+                this.state.googleFontFamily = undefined;
+                const file = this.fileInput.el.files[0];
+                if (!file) {
                     return;
                 }
-
-                const font = m[1].replace(/\+/g, ' ');
-                const googleFontServe = state.googleServe;
-                const fontName = `'${font}'`;
-                // If the font already exists, it will only be added if
-                // the user chooses to add it locally when it is already
-                // imported from the Google Fonts server.
-                const fontExistsLocally = this.googleLocalFonts.some(localFont => localFont.split(':')[0] === fontName);
-                const fontExistsOnServer = this.allFonts.includes(fontName);
-                const preventFontAddition = fontExistsLocally || (fontExistsOnServer && googleFontServe);
-                if (preventFontAddition) {
-                    inputEl.classList.add('is-invalid');
-                    // Show custom validity error message.
-                    inputEl.setCustomValidity(_t("This font already exists, you can only add it as a local font to replace the server version."));
-                    inputEl.reportValidity();
-                    return;
+                const reader = new FileReader();
+                reader.onload = (e) => {
+                    const base64 = e.srcElement.result.split(',')[1];
+                    rpc("/website/theme_upload_font", {
+                        name: file.name,
+                        data: base64,
+                    }).then(result => {
+                        this.state.uploadedFonts = result;
+                        this.updateFontStyle(file.name.substr(0, file.name.lastIndexOf(".")));
+                    });
+                };
+                reader.readAsDataURL(file);
+            }
+            updateFontStyle(baseFontName) {
+                const targetFonts = {};
+                // Add candidate tags to fonts.
+                let shortestNamedFont;
+                for (const font of this.state.uploadedFonts) {
+                    if (!shortestNamedFont || font.name.length < shortestNamedFont.name.length) {
+                        shortestNamedFont = font;
+                    }
+                    font.isItalic = /italic/i.test(font.name);
+                    font.isLight = /light|300/i.test(font.name);
+                    font.isBold = /bold|700/i.test(font.name);
+                    font.isRegular = /regular|400/i.test(font.name);
+                    font.weight = font.isRegular ? 400 : font.isLight ? 300 : font.isBold ? 700 : undefined;
+                    if (font.isItalic && !font.weight) {
+                        if (!/00|thin|medium|black|condense|extrude/i.test(font.name)) {
+                            font.isRegular = true;
+                            font.weight = 400;
+                        }
```

Comment:
The goal was to ignore those fonts on purpose - otherwise they might be used instead of the most basic regular fonts.
This whole matching is very arbitrary - it is a "best effort" in a first version of the feature based on font files we have been able to observe. This might have to be refined if we receive feedback about unsupported files.